### PR TITLE
Add helm-operator

### PR DESCRIPTION
**Pull request by tuananh**
Signed-off-by: Tuan Anh Tran <me@tuananh.org>

## Description

Add `helm-operator` in new format. Courtesy the original work from @aidun here https://github.com/alexellis/arkade/pull/5

## Motivation and Context

Related PR: https://github.com/alexellis/arkade/pull/5

- [ ] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?

- create a new cluster with `kind` `kind create cluster`
- `./arkade install helm-operator`

output

image

check again with `helm list -A`

```
NAME            NAMESPACE               REVISION        UPDATED                                 STATUS          CHART                   APP VERSION
helm-operator   default                 1               2020-10-30 22:11:10.571254 +0700 +07    deployed        helm-operator-1.2.0     1.2.0  
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

- [ ] I have tested this on arm, or have added code to prevent deployment

### CODE REVIEW

File: cmd/apps/helm_operator_app.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		installOptions := types.DefaultInstallOptions().
+			WithNamespace(namespace).
+			WithHelmPath(path.Join(userPath, ".helm")).
+			WithHelmRepo(helmRepo).
+			WithHelmURL("https://charts.fluxcd.io").
+			WithOverrides(overrides).
+			WithWait(wait).
+			WithHelmUpdateRepo(updateRepo)
+
+		if command.Flags().Changed("kubeconfig") {
```
Comment (Waterdrips) [inline cmd/apps/helm_operator_app.go:97]: We should always be getting and setting the kubeconfig flag value and setting it in the options, not only when its changed. 

always grab the value at the start
```go
kubeConfigPath, _ := command.Flags().GetString("kubeconfig")
```
then add the below to the optionas

```go 
			WithKubeconfigPath(kubeConfigPath)

```

**Review by Waterdrips — CHANGES_REQUESTED** [review]

File: cmd/apps/helm_operator_app.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		Use:          "helm-operator",
+		Short:        "Install helm-operator",
+		Long:         "Install helm-operator",
+		Example:      "arkade install helm-operator --namespace default",
+		SilenceUsage: true,
+	}
+
+	helmOperator.Flags().StringP("namespace", "n", "default", "The namespace used for installation")
+	helmOperator.Flags().Bool("update-repo", true, "Update the helm repo")
+	helmOperator.Flags().Bool("helm3", true, "Use helm3, if set to false uses helm2")
```
Comment (Waterdrips) [inline cmd/apps/helm_operator_app.go:33]: just spotted helm2. We no longer support helm2 in arkade, can this be taken out

**Review by Waterdrips — CHANGES_REQUESTED** [review]

File: cmd/apps/helm_operator_app.go (hunk 2 above)
Comment (tuananh) [reply cmd/apps/helm_operator_app.go:33]: Removed

File: cmd/apps/helm_operator_app.go (hunk 1 above)
Comment (tuananh) [reply cmd/apps/helm_operator_app.go]: Done

File: cmd/apps/helm_operator_app.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
```
Comment (Waterdrips) [inline cmd/apps/helm_operator_app.go:33]: need to remove the --helm3 flag still, the code that uses it has gone.

File: cmd/apps/helm_operator_app.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+		helmRepo := "fluxcd/helm-operator"
+
+		err = helm.FetchChart(helmRepo, defaultVersion)
+		if err != nil {
+			return err
+		}
+
+		overrides := map[string]string{}
+		overrides["helm.versions"] = "v3"
```
Comment (Waterdrips) [inline cmd/apps/helm_operator_app.go:77]: This flag, looking at the helm operator readme, states if the helm operator install should allow helm2&3 or just helm3 charts to be installed (by the running operator)

We should either leave it out, or have a flag like `--helm3-only` that's false by default, and not setting this override. 

https://github.com/fluxcd/helm-operator/tree/master/chart/helm-operator#installation

**Review by Waterdrips — CHANGES_REQUESTED** [review]

File: cmd/apps/helm_operator_app.go (hunk 4 above)
Comment (tuananh) [reply cmd/apps/helm_operator_app.go:77]: not sure what I should do here since arkade only supports helm3 from now?

Comment (Waterdrips) [reply cmd/apps/helm_operator_app.go:77]: good point

The readme states 

```
Helm versions supported by this operator instance, if v2 is specified then Tiller is required
```

We should probably just set it to v3, then later on (after merging a user's --set flags) check if they have put something that contains v2 and then print something along the lines of `you have specified helm2 support for the operator, please ensure tiller is installed on your cluster` 

Or you could look at doing a check for the tiller pod before trying to install? 

Im not sure what happens if you try to use `v2` without tiller pre-installed. can you check and come up with a plan?

**Comment by Shikachuu** [comment]
@tuananh, are you still intrested?

**Comment by Shikachuu** [comment]
I think we can assume that this PR has been abandoned by the author.
I'll close this for now, but feel free to revisit this in a new PR!

**Pull request closed without merging**